### PR TITLE
Fix teardown error reporting when `--maxfail=1`

### LINKED ISSUE #11706: Pytest aborts when fixture errors during teardown and `--maxfail=1`
- [x] a detailed description of the bug or problem you are having
- [x] output of `pip list` from the virtual environment you are using
- [x] pytest and operating system versions
- [x] minimal example if possible

## Description

Pytest aborts if a session-scoped fixture errors during teardown when (a) one or more test cases fail and (b) `--maxfail=1` is used.

When this happens, `test-report.xml` does not contain the teardown fixture error. This is problematic as users may want to also record teardown fixture errors when their test cases fail.

### Code

```python
import pytest


def some_condition() -> bool:
    return True


@pytest.fixture(scope="session", autouse=True)
def my_end_of_session_check():
    yield

    print("Running end of session check!")
    if some_condition():
        pytest.fail("Oh noes")


def test_case_1():
    pytest.fail("This is a failing test")


def test_case_2():
     pytest.fail("This is also a failing test")
```

### Args

`pytest example.py --maxfail=1 --junitxml=test-report.xml`

### Output

Console:

```txt
============================================================================== test session starts ===============================================================================
platform darwin -- Python 3.8.13, pytest-7.4.3, pluggy-1.3.0
rootdir: /Users/bbrown/Downloads
collected 2 items

example.py FRunning end of session check!
Traceback (most recent call last):
  File "/Users/bbrown/Downloads/.venv/bin/pytest", line 8, in <module>
    sys.exit(console_main())
  <clipped>
  File "/Users/bbrown/Downloads/.venv/lib/python3.8/site-packages/_pytest/fixtures.py", line 911, in _teardown_yield_fixture
    next(it)
  File "/Users/bbrown/Downloads/example.py", line 14, in my_end_of_session_check
    pytest.fail("Oh noes")
  File "/Users/bbrown/Downloads/.venv/lib/python3.8/site-packages/_pytest/outcomes.py", line 198, in fail
    raise Failed(msg=reason, pytrace=pytrace)
Failed: Oh noes
```

`test-report.xml`:

```xml
<?xml version="1.0" encoding="utf-8"?><testsuites><testsuite name="pytest" errors="0" failures="1" skipped="0" tests="1" time="0.035" timestamp="2023-12-15T08:56:19.980572" hostname="macbookpro"><testcase classname="example" name="test_case_1" time="0.001"><failure message="Failed: This is a failing test">def test_case_1():
&gt;       pytest.fail("This is a failing test")
E       Failed: This is a failing test

example.py:18: Failed</failure></testcase></testsuite></testsuites>
```

### Expected Output

I expect pytest to not abort, and report "1 failed, 1 error, 1 skipped" in both the console and `test-report.xml` when this happens.

### Environment

__OS:__

macOS Sonoma 14.1.2

__Virtual environment:__

```txt
(.venv) % pip list
Package        Version
-------------- -------
exceptiongroup 1.2.0
iniconfig      2.0.0
packaging      23.2
pip            23.3.1
pluggy         1.3.0
pytest         7.4.3
setuptools     56.0.0
tomli          2.0.1
```

### PR DESCRIPTION
**Pull request by bbrown1867**
Closes #11706.

### CODE REVIEW

**Comment by bbrown1867** [comment]
cc: @bluetech

**Review by RonnyPfannschmidt — APPROVED** [review]
this looks good for dealing with a number of edge cases

should we backport it?

**Comment by bbrown1867** [comment]
Thanks for the review @RonnyPfannschmidt! Yeah I think backporting to 7.4.x would be useful to pick these changes up sooner.

**Comment by bbrown1867** [comment]
@RonnyPfannschmidt - is anything needed on my end to get this merged? Or are we waiting for more maintainers to review?

**Comment by RonnyPfannschmidt** [comment]
I'm down as with a headache, so I'd like other's to merge

File: testing/test_runner.py
Comment (bluetech) [inline testing/test_runner.py:1093]:
```suggestion
def test_higher_scoped_teardown_with_maxfail(pytester: Pytester) -> None:
    """Test that higher-scoped fixture teardowns run in the context of the last item
    after the test session bails early due to --maxfail.
    
    Regression test for #11706.
    """
```

File: testing/test_runner.py
Comment (bluetech) [inline testing/test_runner.py:1111]: Better to use [`result.assert_outcomes`](https://docs.pytest.org/en/stable/reference/reference.html#pytest.RunResult.assert_outcomes)

**Review by bluetech — CHANGES_REQUESTED** [review]
Thanks for send a PR @bbrown1867.

I left a couple of comments on the diff, but here are the high level comments:

- Similar to `session.shouldfail`, there is `session.shouldstop`, which we should handle as well. You should be able to test it by using `pytest.exit` instead of `pytest.fail`.

- You've put the `nextitem` check in `runner.py`'s `pytest_runtest_teardown` hook implementation, but I think it should be done in `runtestprotocol`, such that other hook impls in other plugins get `nextitem=None` as well.

- I am still somewhat worried about this being quite subtle and other plugins might mess with with `shouldfail`/`shouldstop` which will then cause very confusing situations. Currently I can't come up with a better fix, so I think we should do it. But I'd like to mitigate the risk a bit: Let's turn `shouldstop` and `shouldfail` into properties, with setters which prevent true -> false state transitions.

As for backporting, I think this is too risky for that. And considering that it is not a regression but has been this way forever AFAICT, I think we shouldn't backport.

**Review by bbrown1867 — COMMENTED** [review]
Thanks for the feedback @bluetech! I think I've implemented all of your feedback.

File: src/_pytest/main.py
Comment (bluetech) [reply src/_pytest/main.py:529]: I think silently ignoring is no good, it will be confusing why there is no effect.

The two options are issuing a warning, or raising an exception. Since we're already stopping raising an exception seems kinda pointless, so I think we should go with a warning. I think it can be something like this:

```py
# The runner checks shouldfail and assumes that if it is set we are definitely stopping,
# so prevent unsetting it.
warnings.warn(PytestWarning("session.shouldfail cannot be unset after it has been set; ignoring."), stacklevel=2)
```

Unfortunately it does mean you'll need to write extra tests to cover these warnings, let me know if you need help with that.

Comment (bbrown1867) [reply src/_pytest/main.py:597]: Okay sounds good, I've added the warnings and and a unit test for the `shouldfail` scenario called `test_shouldfail_is_sticky`. I'm using `pytest_sessionfinish` to trigger the bad behavior. However this same approach did not work for `shouldstop` - it seems like `shouldstop` was still `False` in `pytest_sessionfinish` even if I changed `pytest.fail` to `pytest.exit`. Any ideas on how to test the `shouldstop` scenario?

Comment (bluetech) [reply src/_pytest/main.py:596]: The only thing which sets `shouldstop` in pytest core is `--stepwise`, which actually has the same problem as `--maxfail`. Since I've already checked out the branch, I've pushed the new test along with some tweaks. This is now good to go, thanks!

**Review by bluetech — APPROVED** [review]

**Pull request merged**
